Server/Application/CashByCodeMethods/Command/GenerateCashByCodeCommand.cs

```diff
@@ -0,0 +1,76 @@
+﻿using Endava_Project.Server.Data;
+using Endava_Project.Server.Helpers;
+using Endava_Project.Server.Models;
+using MediatR;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Endava_Project.Server.Application.CashByCodeMethods.Command
+{
+    public class GenerateCashByCodeCommand : IRequest<CommandResult>
+    {
+        public string UserId { get; set; }
+        public Guid SourceWalletId { get; set; }
+        public decimal Amount { get; set; }
+    }
+
+    public class GenerateCashByCodeHandler : IRequestHandler<GenerateCashByCodeCommand, CommandResult>
+    {
+        private readonly ApplicationDbContext context;
+        private readonly IServiceScopeFactory serviceScopeFactory;
+
+        public GenerateCashByCodeHandler(ApplicationDbContext context, IServiceScopeFactory serviceScopeFactory)
+        {
+            this.context = context;
+            this.serviceScopeFactory = serviceScopeFactory;
+        }
+
+        public async Task<CommandResult> Handle(GenerateCashByCodeCommand command , CancellationToken cancellationToken)
+        {
+            int codeLenght = 6;
+
+            var user = await context.Users.Include(x => x.Wallets).FirstOrDefaultAsync(x => x.Id == command.UserId);
+
+            if (!user.Wallets.Any(x => x.Id == command.SourceWalletId))
+                return CommandResult.ReturnFailure();
+
+            var sourceWallet = context.Wallets.FirstOrDefault(e => e.Id == command.SourceWalletId);
+
+            if(sourceWallet.Amount == 0 || sourceWallet.Amount<command.Amount)
+                return CommandResult.ReturnFailure();
+
+            sourceWallet.Amount -= command.Amount;
+
+            var cashByCode = new CashByCode
+            {
+                SourceWalletId = command.SourceWalletId,
+                SourceUserId = command.UserId,
+                Amount = command.Amount,
+                Currency = sourceWallet.Currency,
+                ExpireTime = DateTime.Now.AddMinutes(60),
+                GeneratedCode = StringGenerator.Generator(codeLenght)
+            };
+            context.Add(cashByCode);
+            context.SaveChanges();
+
+            _ = Task.Run(() => LifeSpan(cashByCode.Id, serviceScopeFactory));
+
+            return CommandResult.ReturnSuccess();
+        }
+
+
+        private async Task LifeSpan(Guid Id,IServiceScopeFactory serviceScopeFactory)
```

Comment:
How about we trigger an event when a CashByCode is created?
```csharp
public class CashByCodeCreated : INotification
...
mediator.Publish(new CashByCodeCreated(...));
...
public class CashByCodeCreatedHandler : INotificationHandler<CashByCodeCreated>()
```

something along these lines.

In this way, the logic of creating a cash by code doesn't have unnecessary responsibilities (like deleting it after X seconds). (Single Responsibility Principle)
Also, in this way, we can attach additional  logic when a cash by code is created without modifying this class (like sending an email notification if we really want). (Open-closed Principle)

Reply:
but if this feels too complicated, we can always implement this in another PR

Reply:
complicaterd or not , if i want to become a good .net programmator i need to explore and learn a lot of things .